### PR TITLE
Db2AzureSearch changes for Search-by-TFM

### PR DESCRIPTION
**Pull request by advay26**
Addresses https://github.com/NuGet/Engineering/issues/4575

*NOTE:* We are using a feature branch so that we can stack a few PRs together before merging it into dev. The Catalog2AzureSearch changes will follow this PR.

This change adds framework data to the Search Index through the Db2AzureSearch job:
* Create new index fields for 'Frameworks' and 'Tfms'
* Populate the 'Tfms' field with target framework data from the DB
* Map the same Tfms to their respective Framework generations for the 'Frameworks' field
* Add tests

### CODE REVIEW

File: tests/NuGet.Services.AzureSearch.Tests/HijackDocumentBuilderFacts.cs [hunk 1]
```diff
@@ -299,6 +305,7 @@ public async Task SetsExpectedProperties()
       ""description"": ""Description."",
       ""fileSize"": 3039254,
       ""flattenedDependencies"": ""Microsoft.Data.OData:5.6.4:net40-client|Newtonsoft.Json:6.0.8:net40-client"",
+      ""frameworks"": null,
```
Comment (advay26) [inline tests/NuGet.Services.AzureSearch.Tests/HijackDocumentBuilderFacts.cs:308]: I wanted to split the Db2AzureSearch and Catalog2AzureSearch work into 2 PRs for easier readability. To make this possible, I've had to put the 'frameworks' and 'tfms' fields as null in the Catalog test cases (rather than an array), as Catalog2AzureSearch is not currently processing this data. I will add the Catalog2AzureSearch changes, including tests, in the next PR.

File: src/NuGet.Services.AzureSearch/Models/BaseMetadataDocument.cs [hunk 2]
```diff
@@ -24,6 +24,10 @@ public class BaseMetadataDocument : CommittedDocument, IBaseMetadataDocument
 
         public long? FileSize { get; set; }
         public string FlattenedDependencies { get; set; }
+
+        [SimpleField(IsFilterable = true)]
+        public string[] Frameworks { get; set; }
```
Comment (joelverhagen) [inline src/NuGet.Services.AzureSearch/Models/BaseMetadataDocument.cs:29]: Any particular reason we put this on the BaseMetadataDocument? This document is used for both the hijack index and the search index IIRC. I don't think we need to put this on the hijack index, right?

Comment (joelverhagen) [inline src/NuGet.Services.AzureSearch/Models/BaseMetadataDocument.cs:29]: We should have code comments explaining which part of the TFM the "framework" is.

Comment (joelverhagen) [reply src/NuGet.Services.AzureSearch/Models/BaseMetadataDocument.cs:29]: Also, normalization format, examples, etc.

Comment (joelverhagen) [reply src/NuGet.Services.AzureSearch/Models/BaseMetadataDocument.cs:29]: (use XML docs)

File: src/NuGet.Services.AzureSearch/BaseDocumentBuilder.cs [hunk 3]
```diff
@@ -319,5 +321,45 @@ private static void AddFlattenedFrameworkDependency(NuGetFramework targetFramewo
                 }
             }
         }
+
+        private static string[] GetFrameworksFromPackage(ICollection<PackageFramework> supportedFrameworks)
+        {
+            var tfms = supportedFrameworks
+                            .Where(f => !f.FrameworkName.IsUnsupported)
```
Comment (joelverhagen) [inline src/NuGet.Services.AzureSearch/BaseDocumentBuilder.cs:328]: Probably worth filtering in only `IsSpecificFramework == true`, https://github.com/NuGet/NuGet.Client/blob/1a082949ae5b6da7ca2cce047396c53ae1afdde7/src/NuGet.Core/NuGet.Frameworks/NuGetFramework.cs#L433-L436. Thoughts @dannyjdev ?

File: src/NuGet.Services.AzureSearch/BaseDocumentBuilder.cs [hunk 4]
```diff
@@ -319,5 +321,45 @@ private static void AddFlattenedFrameworkDependency(NuGetFramework targetFramewo
...
+
+        private static string[] ParseFrameworkGenerations(ICollection<NuGetFramework> tfms)
+        {
+            var frameworks = new HashSet<string>();
+            foreach (var framework in tfms)
+            {
+                switch (framework.Framework)
+                {
+                    case ".NETFramework":
+                        frameworks.Add("netframework");
```
Comment (joelverhagen) [inline src/NuGet.Services.AzureSearch/BaseDocumentBuilder.cs:351]: I see what we're going for here. We need a new string for .NET Framework because we treat it differently than .NET 5 and 6 right?

We're essentially inventing a generation moniker that disambiguates .NET 5+ and .NET Framework ("net" can't do this). I see we need this for the data layer (Azure Search documents) as well as the protocol. Keeping these two sets of strings (data layer and protocol) seems reasonable -- although it's not clear at this point if that's your intent since this is just the data layer and service changes are not in PR yet.

I think it's worth running this by Nikolche and Andy on the client team to make sure there is no existing strings to disambiguate between .NET Framework, .NET 5+, .NET Core, and .NET Standard.

Basically, if available, we should use some official string for these things.

Comment (joelverhagen) [reply src/NuGet.Services.AzureSearch/BaseDocumentBuilder.cs:351]: Also, please define string constants for these generation names and use those. They will likely be used in multiple places or are, at the very least, becoming part of our contract and shouldn't be hidden away in a method body.

**Review by joelverhagen — CHANGES_REQUESTED** [review]
Good so far, just have some comments about the generation name string constants.

File: src/NuGet.Services.AzureSearch/BaseDocumentBuilder.cs (hunk 3 above)
Comment (dannyjdev) [reply src/NuGet.Services.AzureSearch/BaseDocumentBuilder.cs]: Yes, I think it's better to have `IsSpecificFramework` instead of `IsUnsupported`. On gallery we are already filtering the `any` TFM.

File: src/NuGet.Services.AzureSearch/Models/BaseMetadataDocument.cs (hunk 2 above)
Comment (advay26) [reply src/NuGet.Services.AzureSearch/Models/BaseMetadataDocument.cs]: I'd got this wrong, I've now put these fields in the SearchDocument

Comment (advay26) [reply src/NuGet.Services.AzureSearch/Models/BaseMetadataDocument.cs]: I've added docs for the 2 new fields (now in SearchDocument.cs)

File: src/NuGet.Services.AzureSearch/BaseDocumentBuilder.cs (hunk 4 above)
Comment (advay26) [reply src/NuGet.Services.AzureSearch/BaseDocumentBuilder.cs]: I've created these constants in NuGetGallery https://github.com/NuGet/NuGetGallery/blob/main/src/NuGetGallery.Core/Services/AssetFrameworkHelper.cs#L116-#L128

File: src/NuGet.Services.AzureSearch/BaseDocumentBuilder.cs [hunk 5]
```diff
@@ -319,5 +321,45 @@ private static void AddFlattenedFrameworkDependency(NuGetFramework targetFramewo
...
+        }
+
+        private static string[] ParseFrameworkGenerations(ICollection<NuGetFramework> tfms)
+        {
+            var frameworks = new HashSet<string>();
+            foreach (var framework in tfms)
+            {
+                switch (framework.Framework)
+                {
+                    case ".NETFramework":
```
[the remark this answers is not in the document]
Comment (advay26) [reply src/NuGet.Services.AzureSearch/BaseDocumentBuilder.cs]: I've changed up this switch case to use available constants where we can, and to use case-insensitive comparisons for these values.

File: src/NuGet.Services.AzureSearch/SearchDocumentBuilder.cs [hunk 6]
```diff
@@ -357,5 +371,47 @@ private static void PopulateKey(KeyedDocument document, string packageId, Search
         {
             document.IsExcludedByDefault = isExcludedByDefault;
         }
+
+        private static string[] GetFrameworksFromPackage(ICollection<PackageFramework> supportedFrameworks)
+        {
+            var tfms = supportedFrameworks
+                            .Where(f => f.FrameworkName.IsSpecificFramework)
```
Comment (dannyjdev) [inline src/NuGet.Services.AzureSearch/SearchDocumentBuilder.cs:378]: Also, filter Portable Class Libraries [PackageFrameworkCompatibility.Create](https://github.com/NuGet/NuGetGallery/blob/main/src/NuGetGallery.Core/Frameworks/PackageFrameworkCompatibilityFactory.cs#L41) (the `!f.IsPCL` part) since we are not going to support PCLs on this work. Make sure to update his on the `GetTfmsFromPackage` method as well.

**Review by joelverhagen — APPROVED** [review]

**Review by dannyjdev — APPROVED** [review]

**Review by joelverhagen — APPROVED** [review]
